wire-ios-data-model/Tests/MLS/MLSServiceTests.swift

```diff
@@ -571,6 +571,87 @@ final class MLSServiceTests: ZMConversationTestsBase, MLSServiceDelegate {
         XCTAssertTrue(mockAddMembersCalled)
     }
 
+    func test_reEstablishGroup_establishGroup() async throws {
+        // GIVEN
+        let groupID = MLSGroupID(Data([1, 2, 3]))
+        let conversation = await uiMOC.perform { self.createConversation(
+            outOfSync: false,
+            currentEpoch: 0,
+            groupID: groupID
+        ).conversation }
+        await uiMOC.perform {
+            XCTAssertEqual(conversation.mlsStatus, .pendingJoin)
+        }
+        let mlsSelfUser = await uiMOC.perform {
+            MLSUser(from: self.selfUser, localDomain: self.localDomain)
+        }
+        let groupInfo = Data()
+        mockActionsProvider
+            .fetchConversationGroupInfoConversationIdDomainSubgroupTypeContext_MockMethod = { _, _, _, _ in
+                groupInfo
+            }
+        mockMLSActionExecutor.mockJoinGroup = { mlsGroupID, mlsGroupInfo in
+            XCTAssertEqual(mlsGroupID, groupID)
+            XCTAssertEqual(mlsGroupInfo, groupInfo)
+        }
+        mockActionsProvider.syncConversationQualifiedIDContext_MockMethod = { _, _ in }
+        mockCoreCryptoContext.conversationExistsConversationId_MockValue = true
+
+        // WHEN
+        try await sut.reEstablishPendingGroup(groupID: groupID)
+
+        // THEN
+        try XCTAssertCount(mockActionsProvider.syncConversationQualifiedIDContext_Invocations, count: 1)
+        try XCTAssertCount(
+            mockActionsProvider.fetchConversationGroupInfoConversationIdDomainSubgroupTypeContext_Invocations,
+            count: 1
+        )
+        await uiMOC.perform {
+            XCTAssertEqual(conversation.mlsStatus, .ready)
+        }
+    }
```

Comment:
This test and the one below are very similar. Only the conversation epoch appears to change. Perhaps the code duplication can be avoided.

I'm guessing we want to test that if the epoch is 0 we establish the group, but if the epoch is >0 we join via external commit ? However with how the test is set up we only verify that the status changes from `pendingJoin` to `ready`. Can we test whether the group was established or joined via external commit ?